File: src/transfers.ts
import type { Exchange } from "ccxt";
import { fetchJson } from "./http";
import { ChainInfo, TransferStatus } from "./models";
import { normalizeChainName } from "./chains";

function coerceInt(value: unknown): number | null {
  const n = Number(value);
  if (!Number.isFinite(n)) return null;
  return Math.trunc(n);
}

type BithumbInoutPayload = {
  data?: Array<{
    coinSymbol?: string;
    networkInfoList?: Array<{
      networkName?: string;
      networkKey?: string;
      isDepositAvailable?: boolean;
      isWithdrawAvailable?: boolean;
      depositConfirmCount?: unknown;
      withdrawFeeQuantity?: unknown;
      withdrawMinimumQuantity?: unknown;
    }>;
  }>;
};

export async function bithumbInoutStatuses(coins: Iterable<string>): Promise<Record<string, TransferStatus>> {
  const coinSet = new Set(Array.from(coins, (c) => (c || "").toUpperCase()).filter(Boolean));
  const out: Record<string, TransferStatus> = {};
  for (const c of coinSet) {
    out[c] = { exchange: "bithumb", coin: c, depositOk: null, withdrawOk: null, chains: [], chainInfo: [] };
  }
  if (coinSet.size === 0) return out;

  try {
    const payload = await fetchJson<BithumbInoutPayload>("https://gw.bithumb.com/exchange/v1/coin-inout/info", {
      timeoutMs: 5000,
    });
    const coinList = payload?.data ?? [];
    const rowMap = new Map<string, any>();
    for (const row of coinList) {
      const sym = (row?.coinSymbol ?? "").toUpperCase();
      if (sym) rowMap.set(sym, row);
    }

    for (const coin of coinSet) {
      const target = rowMap.get(coin);
      if (!target) continue;

      const networks: any[] = target?.networkInfoList ?? [];
      const depositOk = networks.length ? networks.some((n) => n?.isDepositAvailable) : null;
      const withdrawOk = networks.length ? networks.some((n) => n?.isWithdrawAvailable) : null;

      const chains: string[] = [];
      const chainInfo: ChainInfo[] = [];
      for (const net of networks) {
        const name = String(net?.networkName ?? net?.networkKey ?? "NETWORK");
        const depOk = typeof net?.isDepositAvailable === "boolean" ? net.isDepositAvailable : null;
        const wdOk = typeof net?.isWithdrawAvailable === "boolean" ? net.isWithdrawAvailable : null;
        const confirmations = coerceInt(net?.depositConfirmCount);
        const withdrawFee = net?.withdrawFeeQuantity != null ? Number(net.withdrawFeeQuantity) : null;
        const withdrawMin = net?.withdrawMinimumQuantity != null ? Number(net.withdrawMinimumQuantity) : null;
        chainInfo.push({
          name,
          depositOk: depOk,
          withdrawOk: wdOk,
          confirmations,
          withdrawFee: Number.isFinite(withdrawFee) ? withdrawFee : null,
          withdrawMin: Number.isFinite(withdrawMin) ? withdrawMin : null,
        });
        if (depOk === true && wdOk === true) chains.push(name);
      }

      out[coin] = { exchange: "bithumb", coin, depositOk, withdrawOk, chains, chainInfo };
    }
  } catch {
    return out;
  }

  return out;
}

export async function gateioCurrencyStatuses(
  spot: Exchange,
  coins: Iterable<string>,
): Promise<Record<string, TransferStatus>> {
  const coinSet = new Set(Array.from(coins, (c) => (c || "").toUpperCase()).filter(Boolean));
  const out: Record<string, TransferStatus> = {};
  for (const c of coinSet) {
    out[c] = { exchange: "gateio", coin: c, depositOk: null, withdrawOk: null, chains: [], chainInfo: [] };
  }
  if (coinSet.size === 0) return out;

  let currencies: any;
  try {
    currencies = await spot.fetchCurrencies();
  } catch {
    return out;
  }

  let withdrawFees: any = null;
  try {
    if (typeof (spot as any).fetchDepositWithdrawFees === "function") {
      withdrawFees = await (spot as any).fetchDepositWithdrawFees(Array.from(coinSet));
    }
  } catch {
    withdrawFees = null;
  }

  for (const coin of coinSet) {
    const info: any = currencies?.[coin] ?? {};
    const deposit = info?.deposit;
    const withdraw = info?.withdraw;
    const depositOk = typeof deposit === "boolean" ? deposit : deposit != null ? Boolean(deposit) : null;
    const withdrawOk = typeof withdraw === "boolean" ? withdraw : withdraw != null ? Boolean(withdraw) : null;

    const chains: string[] = [];
    const chainInfo: ChainInfo[] = [];
    const networks = info?.networks;
    if (networks && typeof networks === "object") {
      for (const [netCode, netInfo] of Object.entries<any>(networks)) {
        const depOk = typeof netInfo?.deposit === "boolean" ? netInfo.deposit : netInfo?.deposit != null ? Boolean(netInfo.deposit) : null;
        const wdOk = typeof netInfo?.withdraw === "boolean" ? netInfo.withdraw : netInfo?.withdraw != null ? Boolean(netInfo.withdraw) : null;
        let withdrawFee: number | null = null;
        const feeRow = withdrawFees?.[coin];
        const feeNetworks: any = feeRow?.networks;
        if (feeNetworks && typeof feeNetworks === "object") {
          const direct = feeNetworks[String(netCode)]?.withdraw?.fee;
          if (direct != null && Number.isFinite(Number(direct))) withdrawFee = Number(direct);
          if (withdrawFee == null) {
            const wanted = normalizeChainName(String(netCode));
            for (const [feeNetCode, feeNet] of Object.entries<any>(feeNetworks)) {
              if (normalizeChainName(String(feeNetCode)) !== wanted) continue;
              const candidate = feeNet?.withdraw?.fee;
              if (candidate != null && Number.isFinite(Number(candidate))) withdrawFee = Number(candidate);
              break;
            }
          }
        } else if (feeRow?.withdraw?.fee != null && Number.isFinite(Number(feeRow.withdraw.fee))) {
          withdrawFee = Number(feeRow.withdraw.fee);
        }

        chainInfo.push({
          name: String(netCode),
          depositOk: depOk,
          withdrawOk: wdOk,
          confirmations: null,
          withdrawFee,
        });
        if (depOk === true && wdOk === true) chains.push(String(netCode));
      }
    }

    out[coin] = { exchange: "gateio", coin, depositOk, withdrawOk, chains, chainInfo };
  }

  return out;
}

File: legacy_py/arbitrage/transfers.py
"""Deposit/withdraw status helpers."""

from __future__ import annotations

from typing import Any, Dict, Iterable, Optional

import ccxt
import requests

from .models import ChainInfo, TransferStatus


def _coerce_int(value: Any) -> Optional[int]:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def bithumb_inout_status(coin: str) -> TransferStatus:
    """Fetches Bithumb deposit/withdraw status (public endpoint)."""
    url = "https://gw.bithumb.com/exchange/v1/coin-inout/info"
    try:
        resp = requests.get(url, timeout=5)
        resp.raise_for_status()
        payload = resp.json()
        coin_list = payload.get("data") or []
        target = next((c for c in coin_list if (c.get("coinSymbol") or "").upper() == coin.upper()), None)
        if not target:
            return TransferStatus(exchange="bithumb", coin=coin.upper(), deposit_ok=None, withdraw_ok=None, chains=())

        networks = target.get("networkInfoList") or []
        deposit_ok = any(net.get("isDepositAvailable") for net in networks) if networks else None
        withdraw_ok = any(net.get("isWithdrawAvailable") for net in networks) if networks else None

        chains = []
        chain_info = []
        for net in networks:
            name = str(net.get("networkName") or net.get("networkKey") or "NETWORK")
            dep_ok = net.get("isDepositAvailable")
            wd_ok = net.get("isWithdrawAvailable")
            confirmations = _coerce_int(net.get("depositConfirmCount"))
            chain_info.append(ChainInfo(name=name, deposit_ok=dep_ok, withdraw_ok=wd_ok, confirmations=confirmations))

            if dep_ok is True and wd_ok is True:
                chains.append(name)

        return TransferStatus(
            exchange="bithumb",
            coin=coin.upper(),
            deposit_ok=deposit_ok,
            withdraw_ok=withdraw_ok,
            chains=tuple(chains),
            chain_info=tuple(chain_info),
        )
    except Exception:
        return TransferStatus(exchange="bithumb", coin=coin.upper(), deposit_ok=None, withdraw_ok=None, chains=())


def bithumb_inout_statuses(coins: Iterable[str]) -> Dict[str, TransferStatus]:
    """Batch version of `bithumb_inout_status` (single HTTP request)."""
    coin_set = {c.upper() for c in coins if c}
    if not coin_set:
        return {}

    url = "https://gw.bithumb.com/exchange/v1/coin-inout/info"
    out: Dict[str, TransferStatus] = {
        c: TransferStatus(
            exchange="bithumb",
            coin=c,
            deposit_ok=None,
            withdraw_ok=None,
            chains=(),
            chain_info=(),
        )
        for c in coin_set
    }

    try:
        resp = requests.get(url, timeout=5)
        resp.raise_for_status()
        payload = resp.json()
        coin_list = payload.get("data") or []
        row_map: Dict[str, Any] = {}
        for row in coin_list:
            sym = (row.get("coinSymbol") or "").upper()
            if sym:
                row_map[sym] = row

        for coin in coin_set:
            target = row_map.get(coin)
            if not target:
                continue

            networks = target.get("networkInfoList") or []
            deposit_ok = any(net.get("isDepositAvailable") for net in networks) if networks else None
            withdraw_ok = any(net.get("isWithdrawAvailable") for net in networks) if networks else None

            chains: list[str] = []
            chain_info: list[ChainInfo] = []
            for net in networks:
                name = str(net.get("networkName") or net.get("networkKey") or "NETWORK")
                dep_ok = net.get("isDepositAvailable")
                wd_ok = net.get("isWithdrawAvailable")
                confirmations = _coerce_int(net.get("depositConfirmCount"))
                chain_info.append(ChainInfo(name=name, deposit_ok=dep_ok, withdraw_ok=wd_ok, confirmations=confirmations))

                if dep_ok is True and wd_ok is True:
                    chains.append(name)

            out[coin] = TransferStatus(
                exchange="bithumb",
                coin=coin,
                deposit_ok=deposit_ok,
                withdraw_ok=withdraw_ok,
                chains=tuple(chains),
                chain_info=tuple(chain_info),
            )

    except Exception:
        return out

    return out


def gateio_currency_status(spot: ccxt.Exchange, coin: str) -> TransferStatus:
    """Fetches GateIO deposit/withdraw status via ccxt currencies."""
    coin_upper = coin.upper()
    try:
        currencies = spot.fetch_currencies()
    except Exception:
        return TransferStatus(exchange="gateio", coin=coin_upper, deposit_ok=None, withdraw_ok=None, chains=())

    info = currencies.get(coin_upper) or currencies.get(coin_upper.upper()) or {}
    deposit_ok = info.get("deposit")
    withdraw_ok = info.get("withdraw")

    chains: list[str] = []
    networks = info.get("networks")
    if isinstance(networks, dict):
        for net_code, net_info in networks.items():
            if net_info.get("deposit") is True and net_info.get("withdraw") is True:
                chains.append(str(net_code))

    chain_info: list[ChainInfo] = []
    if isinstance(networks, dict):
        for net_code, net_info in networks.items():
            dep_ok = net_info.get("deposit")
            wd_ok = net_info.get("withdraw")
            chain_info.append(
                ChainInfo(name=str(net_code), deposit_ok=dep_ok, withdraw_ok=wd_ok, confirmations=None)
            )

    return TransferStatus(
        exchange="gateio",
        coin=coin_upper,
        deposit_ok=bool(deposit_ok) if deposit_ok is not None else None,
        withdraw_ok=bool(withdraw_ok) if withdraw_ok is not None else None,
        chains=tuple(chains),
        chain_info=tuple(chain_info),
    )


def gateio_currency_statuses(spot: ccxt.Exchange, coins: Iterable[str]) -> Dict[str, TransferStatus]:
    """Batch version of `gateio_currency_status` (single ccxt call)."""
    coin_set = {c.upper() for c in coins if c}
    if not coin_set:
        return {}

    out: Dict[str, TransferStatus] = {
        c: TransferStatus(
            exchange="gateio",
            coin=c,
            deposit_ok=None,
            withdraw_ok=None,
            chains=(),
            chain_info=(),
        )
        for c in coin_set
    }

    try:
        currencies = spot.fetch_currencies()
    except Exception:
        return out

    for coin in coin_set:
        info = currencies.get(coin) or currencies.get(coin.upper()) or {}
        deposit_ok = info.get("deposit")
        withdraw_ok = info.get("withdraw")

        chains: list[str] = []
        chain_info: list[ChainInfo] = []
        networks = info.get("networks")
        if isinstance(networks, dict):
            for net_code, net_info in networks.items():
                name = str(net_code)
                dep_ok = net_info.get("deposit")
                wd_ok = net_info.get("withdraw")
                chain_info.append(ChainInfo(name=name, deposit_ok=dep_ok, withdraw_ok=wd_ok, confirmations=None))
                if dep_ok is True and wd_ok is True:
                    chains.append(name)

        out[coin] = TransferStatus(
            exchange="gateio",
            coin=coin,
            deposit_ok=bool(deposit_ok) if deposit_ok is not None else None,
            withdraw_ok=bool(withdraw_ok) if withdraw_ok is not None else None,
            chains=tuple(chains),
            chain_info=tuple(chain_info),
        )

    return out
